File: lol/src/main/kotlin/com/your/lol/domain/LoLScoreStatsCalculator.java
package com.your.lol.domain;

import static java.util.stream.Collectors.groupingBy;
import static java.util.stream.Collectors.toUnmodifiableList;

import com.your.lol.domain.dto.PreProcessedData;
import com.your.lol.dto.riot.ChallengesDto;
import com.your.lol.dto.riot.MatchDto;
import com.your.lol.dto.riot.ParticipantDto;
import com.your.lol.dto.riot.ParticipantTeamDto;
import com.your.lol.dto.statistics.ChampionDataDto;
import com.your.lol.dto.statistics.StatsDto;
import com.your.lol.support.WebClientFacade;

import java.util.List;
import java.util.Map;
import java.util.Map.Entry;

import lombok.RequiredArgsConstructor;
import org.jetbrains.annotations.NotNull;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
public class LoLScoreStatsCalculator {

    private final WebClientFacade webClientFacade;

    private List<MatchDto> matchDtos;

    public StatsDto calculateStatsBySummonerName(String summonerName) {
        if (matchDtos == null) {
            matchDtos = webClientFacade.requestRiotStatistics(summonerName);
        }
        List<PreProcessedData> preProcessedData = createPreProcessData(summonerName, matchDtos);

        return createStatistics(preProcessedData);
    }

    @NotNull
    private StatsDto createStatistics(List<PreProcessedData> preProcessedData) {
        double winRate = calculateWinRate(preProcessedData);
        Map<String, List<PreProcessedData>> dataByChampions = groupByChampionName(preProcessedData);
        List<ChampionDataDto> championDataDtos = calculateStatsOfChampions(dataByChampions);
        return new StatsDto(winRate, championDataDtos);
    }

    private List<ChampionDataDto> calculateStatsOfChampions(Map<String, List<PreProcessedData>> dataByChampions) {
        return dataByChampions.entrySet().stream()
                .map(this::calculateStatsOfChampion)
                .collect(toUnmodifiableList());
    }

    @NotNull
    private ChampionDataDto calculateStatsOfChampion(Entry<String, List<PreProcessedData>> entry) {
        String championName = entry.getKey();
        List<PreProcessedData> championData = entry.getValue();
        return new ChampionDataDto(
                championName,
                calculateWinRateByChampion(championData),
                championData.size(),
                calculateAverageKda(championData)
        );
    }

    private double calculateWinRateByChampion(List<PreProcessedData> championData) {
        return championData.stream()
                .mapToInt(it -> it.convertTeamWinToNumber())
                .average()
                .getAsDouble();
    }

    private double calculateAverageKda(List<PreProcessedData> championData) {
        long totalGameDuration = championData.stream()
                .mapToLong(it -> it.getGameDuration())
                .sum();

        return championData.stream()
                .mapToDouble(it -> it.getGameDuration() * it.getKda())
                .sum() / totalGameDuration;
    }

    private Map<String, List<PreProcessedData>> groupByChampionName(List<PreProcessedData> preProcessedData) {
        return preProcessedData.stream()
                .collect(groupingBy(PreProcessedData::getChampionName));
    }

    private double calculateWinRate(List<PreProcessedData> preProcessedData) {
        return preProcessedData.stream()
                .mapToInt(PreProcessedData::convertTeamWinToNumber)
                .average()
                .getAsDouble();
    }

    private List<PreProcessedData> createPreProcessData(String summonerName, List<MatchDto> matchDtos) {
        return matchDtos.stream()
                .map(MatchDto::getInfo)
                .map(info -> info.filterAndCreateCompositeDto(summonerName))
                .map(participant -> mapPreProcessData(participant))
                .collect(toUnmodifiableList());
    }

    private PreProcessedData mapPreProcessData(ParticipantTeamDto participantTeam) {
        ParticipantDto participant = participantTeam.getParticipantDto();
        ChallengesDto challenges = participant.getChallenges();
        return PreProcessedData.builder()
                .championName(participant.getChampionName())
                .deathsByEnemyChamps(challenges.getDeathsByEnemyChamps())
                .gameDuration(participant.getGameDuration())
                .kda(challenges.getKda())
                .teamWin(participantTeam.isTeamWin())
                .build();
    }
}

File: lol/src/main/kotlin/com/your/lol/domain/LoLScoreStatsCalculator.kt
package com.your.lol.domain

import com.your.lol.domain.dto.PreProcessedData
import com.your.lol.dto.riot.*
import com.your.lol.dto.statistics.ChampionDataDto
import com.your.lol.dto.statistics.StatsDto
import com.your.lol.support.WebClientFacade
import lombok.RequiredArgsConstructor
import org.springframework.stereotype.Component
import java.util.stream.Collectors

@Component
@RequiredArgsConstructor
class LoLScoreStatsCalculator(
        private val webClientFacade: WebClientFacade) {

    fun calculateStatsBySummonerName(summonerName: String): StatsDto {
        val matchDtos = webClientFacade!!.requestRiotStatistics(summonerName)
        val preProcessedData = createPreProcessData(summonerName, matchDtos)
        return createStatistics(preProcessedData)
    }

    private fun createStatistics(preProcessedData: List<PreProcessedData>): StatsDto {
        val winRate = calculateWinRate(preProcessedData)
        val dataByChampions = groupByChampionName(preProcessedData)
        val championDataDtos = calculateStatsOfChampions(dataByChampions)
        return StatsDto(winRate, championDataDtos)
    }

    private fun calculateStatsOfChampions(dataByChampions: Map<String, List<PreProcessedData>>): List<ChampionDataDto> {
        return dataByChampions.entries.stream()
                .map { entry: Map.Entry<String, List<PreProcessedData>> -> calculateStatsOfChampion(entry) }
                .collect(Collectors.toUnmodifiableList())
    }

    private fun calculateStatsOfChampion(entry: Map.Entry<String, List<PreProcessedData>>): ChampionDataDto {
        val championName = entry.key
        val championData = entry.value
        return ChampionDataDto(
                championName,
                calculateWinRateByChampion(championData),
                championData.size,
                calculateAverageKda(championData)
        )
    }

    private fun calculateWinRateByChampion(championData: List<PreProcessedData>): Double {
        return championData.stream()
                .mapToInt { it: PreProcessedData -> it.convertTeamWinToNumber() }
                .average()
                .asDouble
    }

    private fun calculateAverageKda(championData: List<PreProcessedData>): Double {
        val totalGameDuration = championData.stream()
                .mapToLong { (_, _, _, _, gameDuration): PreProcessedData -> gameDuration }
                .sum()
        return championData.stream()
                .mapToDouble { (_, _, kda, _, gameDuration): PreProcessedData -> gameDuration * kda }
                .sum() / totalGameDuration
    }

    private fun groupByChampionName(preProcessedData: List<PreProcessedData>): Map<String, List<PreProcessedData>> {
        return preProcessedData.stream()
                .collect(Collectors.groupingBy(PreProcessedData::championName))
    }

    private fun calculateWinRate(preProcessedData: List<PreProcessedData>): Double {
        return preProcessedData.stream()
                .mapToInt { obj: PreProcessedData -> obj.convertTeamWinToNumber() }
                .average()
                .asDouble
    }

    private fun createPreProcessData(summonerName: String, matchDtos: List<MatchDto?>): List<PreProcessedData> {
        return matchDtos.stream()
                .map { it?.info }
                .map { it?.filterAndCreateCompositeDto(summonerName) }
                .map { mapPreProcessData(it!!) }
                .collect(Collectors.toUnmodifiableList())
    }

    private fun mapPreProcessData(participantTeam: ParticipantTeamDto): PreProcessedData {
        val participant: ParticipantDto = participantTeam.participantDto
        val challenges: ChallengesDto? = participant.challenges
        return PreProcessedData(
                championName = participant.championName,
                teamId = participant.teamId,
                deathsByEnemyChamps = challenges!!.deathsByEnemyChamps,
                gameDuration = participant.gameDuration,
                kda = challenges!!.kda,
                teamWin = participantTeam.teamWin
        )
    }
}